### PR TITLE
HPCC-28689 Refactor jhtree class structure to support additional node compression methods

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: system/jhtree/ctfile.cpp [hunk 1]
```diff
@@ -1138,6 +1146,18 @@ size32_t CJHTreeBlobNode::getBlobData(unsigned offset, void *dst) const
     return sizeHere;
 }
 
+void CJHTreeRawDataNode::unpack(const void *nodeData, bool needCopy)
+{
+    CJHTreeNode::unpack(nodeData, needCopy);
+    // MORE - what about needCopy flag?
+    const byte *data = ((const byte *) nodeData) + sizeof(hdr);
+    unsigned short len = *reinterpret_cast<const unsigned short *>(data);
```
Comment (ghalliday) [inline system/jhtree/ctfile.cpp:1154]: expanded size is only < 64K.  That seems a bit surprising, but less so when I realised the data isn't compressed.

File: system/jhtree/ctfile.cpp [hunk 2]
```diff
@@ -1078,6 +1106,27 @@ CJHTreeBlobNode::~CJHTreeBlobNode()
...
+    if (keyType & HTREE_COMPRESSED_KEY)
+    {
+        expandedSize = keyHdr->getNodeSize();
+        keyBuf = expandKeys(data, expandedSize);
+    }
+    else
+    {
+        // Do blobs work on indexes that are not compressed - looking at old code I'm not sure they do
+        assertex(!(keyType & COL_PREFIX));
+        expandedSize = hdr.keyBytes + sizeof( __int64 );  // MORE - why is the +sizeof() there?
```
Comment (ghalliday) [inline system/jhtree/ctfile.cpp:1123]: I can't work it out either..  I suspect they don't work.  I think they are already compressed, but from memory the write will not correctly close the lzwcomp - so I think the blobs will be invalid.
Should probably unconditionally assume compressed and fix the writer to close.

**Review by ghalliday — APPROVED** [review]
@richardkchapman only minor comments.  Worth squashing and converting to a real PR.

File: system/jhtree/ctfile.cpp [hunk 3]
```diff
@@ -319,6 +315,7 @@ CWriteNode::~CWriteNode()
 
 bool CWriteNode::add(offset_t pos, const void *indata, size32_t insize, unsigned __int64 sequence)
 {
+    char keyType = keyHdr->getKeyType();
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/ctfile.cpp:318]: getKeyType() is defined as returning char...

File: system/jhtree/ctfile.cpp (hunk 2 above)
Comment (ghalliday) [reply system/jhtree/ctfile.cpp:1123]: will be fixed in a later PR...

**Review by ghalliday — APPROVED** [review]

**Comment by ghalliday** [comment]
@jakesmith please can you sanity check - not to see if more can be done (it can) but to spot any flaws in the refactoring since it is so critical.

File: system/jhtree/ctfile.cpp [hunk 4]
```diff
@@ -1126,6 +1106,27 @@ CJHTreeBlobNode::~CJHTreeBlobNode()
...
+        keyBuf = expandKeys(data, expandedSize);
+    }
+    else
+    {
+        // Do blobs work on indexes that are not compressed - looking at old code I'm not sure they do
+        assertex(!(keyType & COL_PREFIX));
+        expandedSize = hdr.keyBytes + sizeof( __int64 );  // MORE - why is the +sizeof() there?
+        keyBuf = (char *) allocMem(expandedSize);
+        memcpy(keyBuf, data, hdr.keyBytes + sizeof( __int64 ));
+
```
Comment (jakesmith) [inline system/jhtree/ctfile.cpp:1126]: trivial/formatting: extra newline

File: system/jhtree/ctfile.cpp [hunk 5]
```diff
@@ -1126,6 +1106,27 @@ CJHTreeBlobNode::~CJHTreeBlobNode()
...
+        expandedSize = keyHdr->getNodeSize();
+        keyBuf = expandKeys(data, expandedSize);
+    }
+    else
+    {
+        // Do blobs work on indexes that are not compressed - looking at old code I'm not sure they do
+        assertex(!(keyType & COL_PREFIX));
+        expandedSize = hdr.keyBytes + sizeof( __int64 );  // MORE - why is the +sizeof() there?
+        keyBuf = (char *) allocMem(expandedSize);
+        memcpy(keyBuf, data, hdr.keyBytes + sizeof( __int64 ));
```
Comment (jakesmith) [inline system/jhtree/ctfile.cpp:1125]: trivial: could use 'expandedSize' for clarity, i.e.:
```
memcpy(keyBuf, data, expandedSize);
```

File: system/jhtree/ctfile.hpp [hunk 6]
```diff
@@ -197,91 +195,111 @@ class jhtree_decl CKeyHdr : public CWritableKeyNode
...
 
-    void unpack(const void *node, bool needCopy);
-    unsigned __int64 firstSequence;
+    unsigned __int64 firstSequence = 0;
 
     inline size32_t getKeyLen() const { return keyLen; }
-    static char *expandKeys(void *src,size32_t &retsize);
-    static void releaseMem(void *togo, size32_t size);
-    static void *allocMem(size32_t size);
+    virtual void unpack(const void *node, bool needCopy);  // MORE - should move into virtual load()...
```
Comment (jakesmith) [inline system/jhtree/ctfile.hpp:261]: trivial: could have 'override' decaration.

(and other 2 unpack overrides below)

**Review by jakesmith — APPROVED** [review]
@richardkchapman - looks good, some trivial issues.

File: system/jhtree/ctfile.cpp [hunk 7]
```diff
@@ -611,11 +612,6 @@ int CJHTreeNode::locateGT(const char * search, unsigned minIndex) const
     return a;
 }
 
-size32_t CJHTreeNode::getNodeSize() const
-{
-    return keyHdr->getNodeSize();
-}
-
 void CJHTreeNode::unpack(const void *node, bool needCopy)
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/ctfile.cpp:615]: I didn't move so that the compare would be easier

File: system/jhtree/ctfile.cpp (hunk 5 above)
Comment (richardkchapman) [reply system/jhtree/ctfile.cpp:1125]: This code has been deleted in a later PR

File: system/jhtree/ctfile.cpp (hunk 4 above)
Comment (richardkchapman) [reply system/jhtree/ctfile.cpp:1126]: This code has been deleted in a later PR

**Pull request merged**